### PR TITLE
fix(README): do not copy docs/README.md as docs/index.md

### PR DESCRIPTION
**Pull request by kuangp**
## Hey, I just made a Pull Request!

Reviving the abandoned change from https://github.com/backstage/backstage/pull/8541 (cc: @ryanmrodriguez).

I believe the original fix from that change is correct (though unfortunate that it is breaking behavior). My thinking here is that though ensuring `index.md` exists by copying it from root is useful, it is currently not a fully functional feature since we found that mkdocs [edit_uri](https://www.mkdocs.org/user-guide/configuration/#edit_uri) is broken in these scenarios since it will point to a non-existent file. By relying on the native support for `docs/README.md`, mkdocs will instrument edit links correctly. 

Alternatively, if we don't want to go with a breaking change, perhaps the solution should be to expose an option for Backstage integrators to enable/disable this functionality (eg. an option in `app-config.yaml`) since potentially configuring `mkdocs.yml` to rely on essentially a techdocs "hack" doesn't feel correct. 

~Also snuck in a small change to render entity titles in the UI if defined :)~ 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/techdocs/src/search/components/TechDocsSearch.tsx
Comment (iamEAP) [inline plugins/techdocs/src/search/components/TechDocsSearch.tsx:80]: Oooh, can we hold off on this until #10178 is in?  That PR introduces a context for entity metadata that can be picked up in here (rather than potentially having to re-load this metadata over the wire).

**Review by iamEAP — COMMENTED** [review]
> _My thinking here is that though ensuring index.md exists by copying it from root is useful, it is currently not a fully functional feature since we found that mkdocs [edit_uri](https://www.mkdocs.org/user-guide/configuration/#edit_uri) is broken in these scenarios since it will point to a non-existent file. By relying on the native support for docs/README.md, mkdocs will instrument edit links correctly._

Yeah, this makes sense to me.

> _...if we don't want to go with a breaking change, perhaps the solution should be to expose an option for Backstage integrators to enable/disable this functionality..._

The major change is quite unfortunate; it's also a little unclear how exactly to handle, as no one directly depends on `@backstage/plugin-techdocs-node`; they'd instead get hit with the break either via the CLI or in the backend.

I like the config override approach, and treating this more as a bug fix.  Maybe we can use the `legacyUseCaseSensitiveTripletPaths` config as inspiration?  What do you think of this:

- Introduce/document a `legacyCopyReadmeMdToIndexMd` optional boolean config, which conditionally performs the (kinda broken) behavior we have today.  We can default it to `false`, even if it is changing behavior.
- Add/document the relevant flag to the `publish` command in the CLI.
- Update the existing changeset to be `minor` (folks are still likely used to seeing breaking notices in minor changes), highlight the break, but explain the config override to maintain existing behavior.  Also add the CLI and TechDocs backend packages to the changeset.

**Comment by kuangp** [comment]
@iamEAP updated the change as you suggested, let me know if you have any other comments!

**Review by iamEAP — APPROVED** [review]
💯  I just did a quick update to nest the config under `techdocs.generator.mkdocs`. Will merge this in when it passes!

**Comment by ryanmrodriguez** [comment]
@kuangp Thank you for taking the wheel on this. It's been sitting in my todo list. I appreciate your work and I'm glad the use-case spans multiple people. I can finally remove my patch now ;]

**Pull request merged**
